Menagerie of job handler improvements

kofalt opened the pull request:
1. Adding logs to a job will now unconditionally heartbeat that job first. A job consumer now has half the requests over time required when a job is running.
1. Creating a job-completion ticket will prevent that job from being orphaned. A job consumer no longer has to keep heartbeating while an upload is in progress.
1. For this reason, tickets now last an hour instead of 5 minutes.
1. Job tickets are now their own class in `jobs.py`.
1. A bunch of handlers were not using `@directives` or had ineffective `return` statements. Corrected.
1. Moved some repetitive download logic to `util.set_for_download`.
1. Moved some dumb humanizing logic to `util.humanize_validation_error`.
1. Harmonized some, but not all, uses of `self.response.json` variable naming.
1. Slightly reduced the log spam when jobs-next has nothing to return.
1. Fixed a bug in the new job-ticket handler that was not marking jobs as complete / failed.
1. Removed meaningless tempdir warnings. Congrats, CI is 288 lines shorter :tada:

nagem commented:
>Creating a job-completion ticket will prevent that job from being orphaned. A job consumer no longer has to stop heartbeating while an upload is in progress.

Did you mean a job consumer will no longer _have to keep heartbeating_?

nagem commented:
>A bunch of handlers were not using @directives

👍 👍 👍

nagem commented on api/jobs/handlers.py:
```diff
@@ -328,40 +280,30 @@ def next(self):
-        self.response.headers['Content-Disposition'] = 'attachment; filename="config.json"'
-
         # Detect if config is old- or new-style.
         # TODO: remove this logic with a DB upgrade, ref database.py's reserved upgrade section.
```
I believe this db upgrade has been completed, can we remove the check below?

kofalt replied:
I don't know. The instructions in both locations were very clear, so I'm assuming not.

nagem commented on api/jobs/handlers.py:
```diff
@@ -474,18 +414,12 @@ def get_logs_html(self, _id):
-        except Exception: # pylint: disable=broad-except
-            raise APINotFoundException('Job not found')
+        j = Job.get(_id)
+        Queue.mutate(j, {}) # Unconditionally heartbeat
```
The removal of the exception handling causes this endpoint to `500` when the job does not exist rather than `404`. IMO we should be moving towards a state where the API only `500`s on server error, not user/client error. You could switch the `Jobs.get` error to an `APINotFoundExeception` to avoid having to catch Exceptions in handler layers, like we do in container storage classes.

kofalt replied:
I checked down the call stack, and `APINotFoundExeception` is what is thrown in this case. Unless I misread.

nagem replied:
You were right, I misread. I believe it used to be an Exception and I didn't look close enough.

nagem commented on api/jobs/queue.py:
```diff
@@ -417,13 +417,26 @@ def scan_for_orphans():
+
+            # If the job is currently attempting to complete, do not orphan.
+            ticket = JobTicket.find(orphan_candidate['_id'])
+            if ticket is not None and len(ticket) > 0:
```
You could write `if ticket:` and get the same result, but it's alright if you were going for clarity.

kofalt replied:
Yup, I really dislike random-thruthiness checks

nagem commented on api/jobs/queue.py:
```diff
@@ -417,13 +417,26 @@ def scan_for_orphans():
-                    'state': 'running',
-                    'modified': {'$lt': datetime.datetime.utcnow() - datetime.timedelta(seconds=100)},
-                },
+                dict(query, **select),
```
Pretty sneaky sis.

kofalt replied:
:roll_eyes:

nagem commented on api/util.py:
```diff
@@ -174,6 +174,18 @@ def sanitize_string_to_filename(value):
+        key = val_err.relative_path[0]
+    message = val_err.message.replace("u'", "'")
+
+    return 'Gear manifest does not match schema on key ' + key + ': ' + message
```
This should be useful in more than just gear validation errors. If its going to live in general util can we adjust it a bit to not mention gear manifests or take an optional string param that can be inserted into the message response?

kofalt replied:
Mmm, fair, I'll drop the gear specific part

kofalt commented:
Comments addressed, and small log tweak added.

9) Slightly reduced the log spam when jobs-next has nothing to return.

nagem commented on api/jobs/handlers.py:
```diff
@@ -301,67 +259,51 @@ def stats(self):
 
         if job is None:
-            self.abort(400, 'No jobs to process')
+            raise InputValidationException('No jobs to process')
```
Since we're adjusting this anyway, could we use an `APINotFoundException`/`404`? The client didn't make an improperly formatted request.

kofalt replied:
Uh, good question... The route exists and is valid, so a 404 does not feel appropriate here. I wanted to use something that kept the 400, hence my choice. I could make a new error class if you wanted, but it just seemed so trivial I didn't bother initially.

nagem replied:
404 as in a "next job" is missing. IMO an empty response would be the best way to respond but easier to write around a non-200. Not a huge deal we change it.

kofalt replied:
Yeah, the trick here is I didn't want to change behavior for the existing clients. I could go check or change our consumers to not mind it, but it doesn't seem worth a breaking change.

kofalt commented:
10. Fixed a bug in the new job-ticket handler that was not marking jobs as complete / failed.
11. Removed meaningless tempdir warnings. Congrats, CI is 288 lines shorter :tada:

kofalt commented:
10. Fixed a bug in the new job-ticket handler that was not marking jobs as complete / failed.
11. Removed meaningless tempdir warnings. Congrats, CI is 288 lines shorter :tada:

ryansanford commented:
@kofalt @nagem Any final changes or review blocking the merge of this? Correct that there are still no breaking changes here?

I understand this PR to be a blocker to wrapping up the new Flywheel Engine.

nagem commented:
@ryansanford Looks like more changes were added again so I’ll need to review those before it can merge.

kofalt commented:
FWIW, this branch does not need to be a blocker; I have configured e2 to use the new job-complete ticketing that this branch modifies, but I made sure it is very easy to disable. Six character diff.

I can even make that configurable if desired... maybe I should, just in case. :thinking:

ryansanford commented:
I'd prefer this PR to have final review and merge. Thanks for the flexibility though @kofalt

nagem commented:
@ambrussimon do you mind taking a look at the last two commits if you have a chance? They reference your work with job completion tickets.
